### PR TITLE
[READY] Plumbing

### PR DESCRIPTION
**Pull request by Time-Green**
:cl: Time-Green
add: Adds plumbing pipes and some hidden plumbing related machines. They're not available in-game yet. 
sprites: Geyser and geyser pump sprites created by Mey-Ha-Zah!
/:cl:
Adds the following objects:

- Fluid ducts . Like atmos pipe, but for reagents. They're smart and can be seperated by layer and color

- Geysers. They spawn on lavaland and can be harvested with liquid pumps or just beakers. 

- Liquid pumps. For pumping geysers. They need to be on a wire node and connected to a duct network. 

- Plumbable reagent dispenser subtype.

- New plumbing RPD specific RPD.

- Plungers. The reinforced plunger can be used for plunging geysers to activate them. They currently serve no other function and can't be worn as a hat yet.

They're all disabled and admin only.

Adds the following systems:

- Ductnet datum. They pretty much say "hey, we're connected"

- Plumbing component. You can turn any movable atom into a plumbable thing. Comes with overlays and everything. They're essential for everything that has input and outputs, except pipes. That means you can have machines connected to both atmospheric pipes and fluid ducts. They only connect with layer 3 ducts. 

- Fluid subsystem that handles the processing on the plumbing machinery and other plumbing related stuff. 

**Basically atmospherics but with reagents. Reagents are pressurized and thus transferred instantly, to avoid those stupid leftovers in pipes no one likes. I am PRing it now because if I were to add a dozen or so associated machines, it would be too much of a pain to manage/review in one PR. It would also be a massive waste if people didn't even want it in the first place.**

**Also yes all of this works (as far as I know)**

Things I plan to add in later PR's 

1. ~~Layers / colors (maybe)~~ done
2. Machinery (Stuff like chemical processors, filters, special mixing chambers and factories)
3. Make geysers more interesting, so chemists/atmostechs/cargotechs/assistants can go to lavaland and get some exotic stuff.
4. Pipe fractures. Damaged pipes could do the cartooney water soaker beam coming out to shoot at people cause why the fuck not. 

Code concerns:
- Connecting is handled through pipes only. Not sure if this is the best approach. This means plumbing devices cant directly connect to each other without pipes. I could add some exceptions to handle it. I could also take connecting away from pipes and make it a global proc, with alot of exceptions since it'd be handling both plumbing machinery and the ducts.

- Reagent transferring. Basically the transfer is instantaneous from A to B. I did this because stuff getting stuck in pipes could pose some serious mixing problems with chemistry, since it's alot more delicate than atmos. I could understand if people would want it to be like atmos. It wouldn't be difficult to change.

Also for testing I suggest you set the pipe_dispenser to category 3 to print fluid ducts. My tests were done with 
/obj/structure/geyser
/obj/structure/reagent_dispensers/plumbed/storage
/obj/item/plunger/reinforced
/obj/machinery/power/liquid_pump
/obj/item/pipe_dispenser/plumbing
that and standard tools for setting things up

### CODE REVIEW

**Comment by nemvar** [comment]
Sanitation technicians now

**Comment by Fikou** [comment]
make the janitor a plumber NOW

**Comment by ondrej008** [comment]
So, fluid physics and actual fluid spills? Or just fluids in pipes and machines?

**Comment by Time-Green** [comment]
@ondrej008 Just fluids in pipes and machines. I do like tsunami station but it'll have to wait

**Comment by Time-Green** [comment]
image
There's an invisible pump on the geyser

**Comment by habader** [comment]
very nice

**Comment by XDTM** [comment]
great potential, could lead to the introduction of "resource" reagents on top of mineral materials

**Comment by Tlaltecuhtli** [comment]
finally a pipe cleaners update

**Comment by Teepeepee** [comment]
change janitor jumpsuit to mario uniform and if there's a second jannie slot make him luigi uniform

**Comment by CPTANT** [comment]
What happens when stuff reacts in a pipe?

**Comment by XDTM** [comment]
since travel's instant i suppose that won't be an issue

**Comment by CPTANT** [comment]
I still see !!!FUN!!! things happening when someone turns all the water tanks in a system into bombs by pumping some potassium into it.

On the other hand it being a complete hugbox would also be extremely boring.

**Comment by nicbn** [comment]
Could you please make it so that atmos pipes work as pumbling as well? It'd be way nicer and wouldn't require new items and all.

**Comment by Time-Green** [comment]
> Could you please make it so that atmos pipes work as pumbling as well? It'd be way nicer and wouldn't require new items and all.

They work very differently and it would be visually confusing

**Comment by nicbn** [comment]
Well, at least make it so that you can't put a plumbing pipe ontop of a atmos pipe facing the same direction in the same layer (and vice-versa of course) if you're keeping the sprite width the same as atmos pipe width (otherwise make it so that the narrower pipe is on the top), that will make some setups harder but require them to be a little neater.

Also you gotta work on the sprites. If you made the pipes somewhat bigger you could say they are also sewage pipes and make them transfer items (maybe only small or tiny ones) (oh yeah, functional toilets that can flush down items!).

I think that ideally, atmos and fluids should work the same way inside tubes, that way one could, for example, flood the station with water by putting shitloads of water in the distro (of course, that would require fluid physics).

**Comment by HomocidlePancakes** [comment]
Now we can have functional toilets and thus pee and poo. Magnificent.

**Comment by optimumtact** [comment]
@Time-Green can you make placement work like the smart cables please.

**Comment by optimumtact** [comment]
PS the joining of the plumbing to sinks/sources looks a bit janky

**Comment by bgobandit** [comment]
a) thank you, I have wanted to code this for years but have never had the time
b) when this does get completed, plumber should be its own job, it has been fucking ages since we had a new job

**Comment by GuyonBroadway** [comment]
You could make maintainence it's own department and have plumbers and janitors be it's founding members. All hail Maintannica, land of trash and poo pipes!

**Comment by Time-Green** [comment]
@optimumtact Auto-connect isnt a thing yet. I’m pretty sure I can make that work

**Comment by PKPenguin321** [comment]
Love how modular this is, I recall this almost becoming a thing way back in a PR by I think xhuis(?)
IIRC that PR had a central water tank type thing that would replicate water and send it to showers, sinks, etc, but if you put enough of a different reagent into it it would replicate the new reagent instead. Might be getting ahead a bit since this PR seems rather preliminary, but something like that might be cool.

**Comment by fludd12** [comment]
I'm excited to see where this goes, especially since I believe there is going to eventually be a project to unify materials, reagents, and gasses into using one material datum system.

**Comment by FantasticFwoosh** [comment]
I just want a emergency pipe sprinker system for the SM

**Comment by Time-Green** [comment]
image

**Comment by optimumtact** [comment]
it's, it's, it's beautiful!

**Comment by MCterra10** [comment]
@optimumtact I really don't think pipes should autoconnect. Although I am an avid smart wire hater, I can see the use. But atmos pipes and these fluid pipes essentially RELY on not autoconnecting to make fancy setups to make rare gasses possible.

**Comment by optimumtact** [comment]
layering is done by colour sweety

**Comment by nicbn** [comment]
I think the problem with autoconnecting is that while cables FEEL like they should autoconnect (they're flexible and all), pipes don't feel that way, they are rigid, it doesn't seem natural for them to autoconnect.

Also pipe construction isn't awkward like cable construction used to be, so I really feel like pipes shouldn't autoconnect.

**Comment by optimumtact** [comment]
Are you talking about atmos right now?

**Comment by MCterra10** [comment]
I think his comment could refer to both atmos and fluid pipes, and I'm not sure what point you were trying to make anyway. Wires (somewhat) work with autoconnect because there is only one type, and most of the time you're connecting the whole thing to the powernet anyway. Whereas with rigid pipes there are multiple different components like manifolds, pumps, gates, filters and many others, and not only would autoconnect be hard to implement, it would make creating complex setups "difficult" to say the least. Pipes like atmos and fluid pipes also carry distinct gasses/reagents, and you almost always don't want them to connect. Think distro and waste pipes, or (if this gets fleshed out in the future) sewerage and water supply pipes. However cables only carry one thing, and that is power, so with mixing the only real issue is hotwired powernets, but that's not a debate for here.

Basically having pipes link, whether atmos or fluid is a bad idea

**Comment by optimumtact** [comment]
I disagree, the limitations are exactly what drives creative thinking when making complex layouts.

**Comment by nicbn** [comment]
The point I'm trying to make is that autoconnection feels like "bending the tube". It feels fine for power cables because they are flexible, but it feels odd on hard pipes because they can't just bend around to link on one another. The current mechanism behind atmos pipes fits the feeling of hard, non-bendable pipes.

I think that if you want the reagent pipes to use autoconnecting, they should look bendable at least (by using the flexible pipe sprites for example)

**Comment by MCterra10** [comment]
To quote you yourself; "it's not complexity, it's clunkiness." Forcing people to create designs that work around artificially imposed restrictions is one of the worst ideas I've heard in a long time. Kinda reminds of battlefront when ubisoft locked Darth away behind a ridiculous amount of game time to "promote a satisfaction for unlocking him".

**Comment by optimumtact** [comment]
Im fairly certain that was in relation to clunkiness which is implied by artificial ui limitations, such as when tasers had a trigger cooldown.

Im not convinced that this meets that test, the system reduces clunkiness because you have less fiddling and pixel hunting to do, while still allowing complex layouts to be expressed with the use of multiple layers.

Having a maximum layer size encourages some creative thinking, and shouldn't present too high a barrier that complex work can not be done.

@nicbn fair point regarding the sprites. However in general I think anything that refers to the current atmospherics /disposals pipe systems as good examples for anything are probably wrongheaded.

There's a reason the RPD exists and it's because laying pipe isn't fun, at all. It's clunky, fiddly and hard, so most people don't even bother.


Smarter autojoining networks in general are meant to increase the amount of new pipe/wire/plumbing being layed because the investment in time is lower. While still allowing the complexity of intricate designs with the use of layers that don't join across each other.


Overall I believe the system will be much more intuitive and hopefully lead people into more creative solutions as they flex the limits of having a maximum number of layers.

**Comment by GuyonBroadway** [comment]
Just gonna say a fair few atmos setups often depend on two pipes running side by side without the gasses within mixinh and the current layer manifold system doesn't allow for this.

You'll need to expand the amount of pipe layers and add nonmixing manifolds or else double the size of atmos to make enough room to stop pipes autoconnecting and oh god the waste loop has mixed with the distro again!

**Comment by Time-Green** [comment]
> Just gonna say a fair few atmos setups often depend on two pipes running side by side without the gasses within mixinh and the current layer manifold system doesn't allow for this.
> 
> You'll need to expand the amount of pipe layers and add nonmixing manifolds or else double the size of atmos to make enough room to stop pipes autoconnecting and oh god the waste loop has mixed with the distro again!

Just use different colours if pipes, they don’t automaticly connect. They only connect when mousedropping with a wrench

**Comment by optimumtact** [comment]
also, nobody's agree to tackle atmos for that system yet anyway, so it's oftopic

**Comment by MCterra10** [comment]
@optimumtact can I do the honors of calling you bad?

**Comment by optimumtact** [comment]
well if you think it will make a difference

**Comment by MCterra10** [comment]
Orange man bad

**Comment by Detective-Google** [comment]
does this mean we can have the pool PR back? I'd love to have it connected to a liquid network... maybe with the sprinklers PR as well, I can already imagine synthflesh fire sprayers... (or acid.)

also, I really like the idea of having a 'liquid replicator' that uses a *ton* of power to run, in the range of a MW or a few, maybe have the replicator locked under techwebs?

Also, geysers seem like a neat idea, but what's the intention for getting geyser liquids *back on station* from lavaland? right now it seems like the only viable option is to scoop it up with a beaker and carry it, maybe a pipe network from geyser to shuttle, to a holding tank on the shuttle, and then out to the station network could work with the current system? unsure.

**Comment by Time-Green** [comment]
> does this mean we can have the pool PR back? I'd love to have it connected to a liquid network... maybe with the sprinklers PR as well, I can already imagine synthflesh fire sprayers... (or acid.)
> 
> also, I really like the idea of having a 'liquid replicator' that uses a _ton_ of power to run, in the range of a MW or a few, maybe have the replicator locked under techwebs?
> 
> Also, geysers seem like a neat idea, but what's the intention for getting geyser liquids _back on station_ from lavaland? right now it seems like the only viable option is to scoop it up with a beaker and carry it, maybe a pipe network from geyser to shuttle, to a holding tank on the shuttle, and then out to the station network could work with the current system? unsure.

Pool and sprinklers would all work with the plumbing. 

I don't really have any plans to do a massive plumbing thing where all is connected, but it could be done. The code only runs when something is being used so it'd be cheap on processing (i think)

Also about taking fluids from lavaland to the station, I was contemplating between liquid barrel station, the holding tank on the shuttle or like a bluespace fluid transporter thing. Either way it'll be another PR

**Comment by Detective-Google** [comment]
> 
> Pool and sprinklers would all work with the plumbing.
> 
> I don't really have any plans to do a massive plumbing thing where all is connected, but it could be done. The code only runs when something is being used so it'd be cheap on processing (i think)
> 
> Also about taking fluids from lavaland to the station, I was contemplating between liquid barrel station, the holding tank on the shuttle or like a bluespace fluid transporter thing. Either way it'll be another PR

Bluespace transporter could be neat, would need some kind of balancing in terms of cost, via either energy use or build cost probably, but would overall be much easier to use in game

**Comment by Time-Green** [comment]
It's ready for review

The basis for plumbing is done. In the last few commits I added 5 different layers and gave them colors. They are smart and connect to ducts of the same color and layer. Cross-color connecting can be done by drag and dropping the ducts while holding a wrench. Cross-layer connecting is done by a layer manifold. I also modularized some code and fixed some bugs.  

There's still pieces left out, but I'll throw that in another PR.

**Comment by optimumtact** [comment]
Wait, im a bit confused, do you have colour and layers right now?

I was imagining that each colour *is* one of the layers

**Comment by Time-Green** [comment]
Oh, no the colors and layers are independent. Every layer does not have a set color.

**Comment by FantasticFwoosh** [comment]
> Oh, no the colors and layers are independent. Every layer does not have a set color.

So absolutely the same as the current layering system in atmos pipes where the pipes are just identically functioned colored variants of one another. That was kind of cringe :tangerine: -man 

> Also, geysers seem like a neat idea, but what's the intention for getting geyser liquids back on station from lavaland?

#30863 bluespace pipe/pumps obviously unless miners had a special extraction kit rather than adding it back into code to abuse. Geysers should also operate for the Turbine but that's my opinion on how to set up easy natural-energy engines.

**Comment by optimumtact** [comment]
how do you feel about not having layers and colours and just having colours?

**Comment by Time-Green** [comment]
It would get really tedious if you couldnt put more than one duct on a tile. It would also have been a huge waste of my time

**Comment by optimumtact** [comment]
okay

**Comment by GuyonBroadway** [comment]
Having selectivly nonconnecting colours AND nonconnecting layers is actually really really neat as you can set up really compact junctions and stuff.

**Comment by optimumtact** [comment]
Yeah, that was always the long term place I saw all the networks going, given it gives you an N^M level of options for layouts. I just didn't realise they'd already done it and I wanted to try the limited versions first.

However they've done the work so that's fine.

**Comment by nicbn** [comment]
Could you please use the sprites from [this](https://github.com/tgstation/tgstation/blob/58d284574385bb03615ad5028bfa20eb27d4c0dd/icons/obj/tubing.dmi) file? (Removed in #33121)

**Comment by optimumtact** [comment]
oof ouch a conflict

**Comment by Time-Green** [comment]
conflict dead
